source-postgres-batch/driver.go

```diff
@@ -143,51 +149,204 @@ func (drv *BatchSQLDriver) Discover(ctx context.Context, req *pc.Request_Discove
+		return nil, fmt.Errorf("error listing tables: %w", err)
+	}
+	keys, err := discoverPrimaryKeys(ctx, db)
+	if err != nil {
+		return nil, fmt.Errorf("error listing primary keys: %w", err)
 	}
-	defer rows.Close()
 
-	var bindings []*pc.Response_Discovered_Binding
-	for rows.Next() {
-		var tableSchema, tableName, tableType string
-		if err := rows.Scan(&tableSchema, &tableName, &tableType); err != nil {
-			return nil, fmt.Errorf("error scanning result row: %w", err)
-		}
+	// We rebuild this map even though `discoverPrimaryKeys` already built and
+	// discarded it, in order to keep the `discoverTables` / `discoverPrimaryKeys`
+	// interface as stupidly simple as possible, which ought to make it just that
+	// little bit easier to do this generically for multiple databases.
+	var keysByTable = make(map[string]*discoveredPrimaryKey)
+	for _, key := range keys {
+		keysByTable[key.Schema+"."+key.Table] = key
+	}
 
+	var bindings []*pc.Response_Discovered_Binding
+	for _, table := range tables {
 		// Exclude tables in "system schemas" such as information_schema or pg_catalog.
-		if slices.Contains(excludedSystemSchemas, tableSchema) {
+		if slices.Contains(excludedSystemSchemas, table.Schema) {
 			continue
 		}
 
-		var recommendedName = recommendedCatalogName(tableSchema, tableName)
-		var res, err = drv.GenerateResource(recommendedName, tableSchema, tableName, tableType)
+		var tableID = table.Schema + "." + table.Name
+
+		var recommendedName = recommendedCatalogName(table.Schema, table.Name)
+		var res, err = drv.GenerateResource(recommendedName, table.Schema, table.Name, table.Type)
 		if err != nil {
 			log.WithFields(log.Fields{
 				"reason": err,
-				"schema": tableSchema,
-				"table":  tableName,
-				"type":   tableType,
-			}).Warn("unable to generate resource spec for entity")
+				"table":  tableID,
+				"type":   table.Type,
+			}).Warn("unable to generate resource spec")
 			continue
 		}
 		resourceConfigJSON, err := json.Marshal(res)
 		if err != nil {
 			return nil, fmt.Errorf("error serializing resource spec: %w", err)
 		}
 
+		// Try to generate a useful collection schema, but on error fall back to the
+		// minimal schema with the default key [/_meta/polled, /_meta/index].
+		var collectionSchema json.RawMessage
+		var collectionKey []string
+		if tableKey, ok := keysByTable[tableID]; ok {
+			collectionSchema, err = generateCollectionSchema(tableKey.Columns, tableKey.ColumnTypes)
+			if err != nil {
+				log.WithFields(log.Fields{"table": tableID, "err": err}).Warn("unable to generate collection schema")
+			} else {
+				for _, colName := range tableKey.Columns {
+					collectionKey = append(collectionKey, primaryKeyToCollectionKey(colName))
+				}
+			}
+		}
+		if collectionSchema == nil {
+			collectionSchema, err = generateCollectionSchema(nil, nil)
+			if err != nil {
+				log.WithFields(log.Fields{"table": tableID, "err": err}).Warn("unable to generate fallback schema")
```

Comment:
The behavior of `generateCollectionSchema(nil, nil)` reduces to what `generateMinimalSchema()` used to be, so this is effectively that, it just handles failure with a warn-and-continue rather than a panic at startup.

I think I can tweak the code to make that clearer, just a moment.